In [10]:
import pandas as pd

<h1>How do I read and write tabular data?</h1>

<img class="align-center" src="https://pandas.pydata.org/docs/_images/02_io_readwrite.svg">

### <h4>I want to analyze the Titanic passenger data, available as a CSV file.</h4>

In [11]:
titanic = pd.read_csv("titanic.csv")

<p>pandas provides the <a href="https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html#pandas.read_csv" title="pandas.read_csv"><code>read_csv()</code></a> function to read data stored as a csv file into a pandas <code>DataFrame</code>. Pandas supports many different file formats or data sources out of the box (csv, excel, sql, json, parquet, …), each of them with the prefix <code>read_*</code>.</p>

<p>Make sure to always have a check on the data after reading in the data.
When displaying a <code>DataFrame</code>, the first and last 5 rows will be shown by default:</p>

In [12]:
titanic

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


### <h4>I want to see the first 8 rows of a pandas DataFrame.</h4>

In [13]:
titanic.head(8)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S


<p>To see the first N rows of a <code>DataFrame</code>, use the <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html#pandas.DataFrame.head" title="pandas.DataFrame.head"><code>head()</code></a> method with
the required number of rows (in this case 8) as argument.</p>
    
<div class="alert alert-block alert-info">
<p>Note</p>
<p>Interested in the last N rows instead? pandas also provides a
<a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.tail.html#pandas.DataFrame.tail" title="pandas.DataFrame.tail"><code>tail()</code></a> method. For example, <code>titanic.tail(10)</code> will return the last 10 rows of the DataFrame.</p>
</div>

<p>A check on how pandas interpreted each of the column data types can be
done by requesting the pandas <code>dtypes</code> attribute:</p>

In [14]:
titanic.dtypes

,0
PassengerId,int64
Survived,int64
Pclass,int64
Name,object
Sex,object
Age,float64
SibSp,int64
Parch,int64
Ticket,object
Fare,float64


<p>For each of the columns, the used data type is enlisted. The data types
in this <code>DataFrame</code> are integers (<code>int64</code>), floats (<code>float64</code>) and strings (<code>object</code>).</p>

<div class="alert alert-block alert-info">
<p>Note</p>
<p>When asking for the <code>dtypes</code>, no brackets are used!
<code>dtypes</code> is an attribute of a <code>DataFrame</code> and <code>Series</code>. Attributes of a <code>DataFrame</code> or <code>Series</code> do not need brackets. Attributes represent a characteristic of a <code>DataFrame</code>/<code>Series</code>, whereas methods (which require brackets) <em>do</em> something with the
<code>DataFrame</code>/<code>Series</code> as introduced in the <a href="https://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html#min-tut-01-tableoriented">first tutorial</a>.</p>
</div>

### <h4>My colleague requested the Titanic data as a spreadsheet.</h4>

In [15]:
titanic.to_excel("titanic.xlsx", sheet_name="passengers", index=False)

<p>Whereas <code>read_*</code> functions are used to read data to pandas, the <code>to_*</code> methods are used to store data. The <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_excel.html#pandas.DataFrame.to_excel" title="pandas.DataFrame.to_excel"><code>to_excel()</code></a> method stores the data as an excel file.
In the example here, the <code>sheet_name</code> is named <em>passengers</em> instead of the default <em>Sheet1</em>.
By setting <code>index=False</code> the row index labels are not saved in the spreadsheet.</p>

<p>The equivalent read function <code>read_excel()</code> will reload the data to a <code>DataFrame</code>:</p>

In [16]:
titanic = pd.read_excel("titanic.xlsx", sheet_name="passengers")

In [17]:
titanic.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


### <h4>I’m interested in a technical summary of a <code>DataFrame</code></h4>

In [18]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


<p>The method <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html#pandas.DataFrame.info" title="pandas.DataFrame.info"><code>info()</code></a> provides technical information about a
<code>DataFrame</code>, so let’s explain the output in more detail:</p>

<ul>

<li><p>It is indeed a <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="pandas.DataFrame"><code>DataFrame</code></a>.</p></li>

<li><p>There are 891 entries, i.e.&nbsp;891 rows.</p></li>

<li><p>Each row has a row label (aka the <code>index</code>) with values ranging from 0 to 890.</p></li>

<li><p>The table has 12 columns. Most columns have a value for each of the
rows (all 891 values are <code>non-null</code>). Some columns do have missing values and less than 891 <code>non-null</code> values.</p></li>

<li><p>The columns <code>Name</code>, <code>Sex</code>, <code>Cabin</code> and <code>Embarked</code> consists of textual data (strings, aka <code>object</code>). The other columns are numerical data with some of them whole numbers (aka <code>integer</code>) and others are real numbers (aka <code>float</code>).</p></li>

<li><p>The kind of data (characters, integers,…) in the different columns
are summarized by listing the <code>dtypes</code>.</p></li>

<li><p>The approximate amount of RAM used to hold the DataFrame is provided
as well.</p></li>

</ul>

<div>
<h4>REMEMBER</h4><ul>
<li><p>Getting data in to pandas from many different file formats or data
sources is supported by <code>read_*</code> functions.</p></li>
<li><p>Exporting data out of pandas is provided by different
<code>to_*</code>methods.</p></li>
<li><p>The <code>head</code>/<code>tail</code>/<code>info</code> methods and the <code>dtypes</code> attribute
are convenient for a first check.</p></li>
</ul>
</div>

<div>
<p><b>To user guide</b> For a complete overview of the input and output possibilities from and to pandas, see the user guide section about <a href="https://pandas.pydata.org/docs/user_guide/io.html#io">reader and writer functions</a>.</p>
</div>